### ISSUE TITLE
lab06

### ISSUE THREAD

**Issue by nikhilreddy06228** [nikhilreddy06228/lab06#1]
# -----------------------------
# Lab06: Decision Tree Analysis
# Dataset: Students Performance
# Tasks A1 - A7
# -----------------------------

import pandas as pd
import numpy as np
from collections import Counter
from math import log2
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.preprocessing import LabelEncoder

# -----------------------------
# Load Dataset
# -----------------------------
file_path = "Students Performance Dataset.csv"  # Update path if needed
df = pd.read_csv(file_path)

# -----------------------------
# A1: Entropy Function
# -----------------------------
def entropy(y):
    counts = Counter(y)
    total = len(y)
    return -sum((count/total) * log2(count/total) for count in counts.values())

# -----------------------------
# A2: Gini Index Function
# -----------------------------
def gini_index(y):
    counts = Counter(y)
    total = len(y)
    return 1 - sum((count/total)**2 for count in counts.values())

# -----------------------------
# A3: Information Gain
# -----------------------------
def information_gain(X_col, y):
    total_entropy = entropy(y)
    values, counts = np.unique(X_col, return_counts=True)
    weighted_entropy = sum((counts[i]/len(X_col)) * entropy(y[X_col == v]) 
                           for i, v in enumerate(values))
    return total_entropy - weighted_entropy

# -----------------------------
# A4: Equal Width Binning
# -----------------------------
def equal_width_binning(feature, bins=4):
    return pd.cut(feature, bins=bins, labels=False)

# -----------------------------
# Encode categorical variables
# -----------------------------
df_copy = df.copy()
label_encoders = {}
for col in df_copy.select_dtypes(include=["object"]).columns:
    if col not in ["Student_ID", "First_Name", "Last_Name", "Email"]:
        le = LabelEncoder()
        df_copy[col] = le.fit_transform(df_copy[col])
        label_encoders[col] = le

# -----------------------------
# Features & Target
# -----------------------------
X = df_copy.drop(columns=["Grade", "Student_ID", "First_Name", "Last_Name", "Email"])
y = df_copy["Grade"]

# -----------------------------
# A1 & A2: Compute Entropy & Gini
# -----------------------------
entropy_value = entropy(y)
gini_value = gini_index(y)

print("A1: Entropy of Grade =", entropy_value)
print("A2: Gini Index of Grade =", gini_value)

# -----------------------------
# A3: Information Gain for each feature
# -----------------------------
ig_scores = {col: information_gain(X[col], y) for col in X.columns}
best_root = max(ig_scores, key=ig_scores.get)

print("\nA3: Information Gain Scores:")
for k, v in ig_scores.items():
    print(f"{k}: {v:.4f}")
print("Best Root Attribute:", best_root)

# -----------------------------
# A4: Apply Equal Width Binning
# -----------------------------
binned_final = equal_width_binning(df["Final_Score"], bins=4)
print("\nA4: Equal-width binning of Final_Score (first 10):", binned_final.head(10).tolist())

# -----------------------------
# A5: Build Decision Tree Classifier
# -----------------------------
clf = DecisionTreeClassifier(criterion="entropy", max_depth=4, random_state=42)
clf.fit(X, y)
print("\nA5: Decision Tree Classifier built successfully!")

# -----------------------------
# A6: Visualize Decision Tree
# -----------------------------
plt.figure(figsize=(15, 8))
plot_tree(clf, filled=True, feature_names=X.columns, 
          class_names=label_encoders["Grade"].classes_)
plt.title("Decision Tree Visualization (A6)")
plt.show()

# -----------------------------
# A7: Decision Boundary with 2 Features
# -----------------------------
def plot_decision_boundary(X, y, clf, feature_names):
    x_min, x_max = X[:, 0].min() - 1, X[:, 0].max() + 1
    y_min, y_max = X[:, 1].min() - 1, X[:, 1].max() + 1
    xx, yy = np.meshgrid(np.arange(x_min, x_max, 0.2),
                         np.arange(y_min, y_max, 0.2))
    
    Z = clf.predict(np.c_[xx.ravel(), yy.ravel()])
    Z = Z.reshape(xx.shape)
    
    plt.contourf(xx, yy, Z, alpha=0.4)
    plt.scatter(X[:, 0], X[:, 1], c=y, edgecolors="k")
    plt.xlabel(feature_names[0])
    plt.ylabel(feature_names[1])
    plt.title("Decision Boundary (A7)")
    plt.show()

feat1, feat2 = "Final_Score", "Study_Hours_per_Week"
X_two = df_copy[[feat1, feat2]].values
clf_two = DecisionTreeClassifier(criterion="entropy", max_depth=3, random_state=42)
clf_two.fit(X_two, y)

plot_decision_boundary(X_two, y, clf_two, [feat1, feat2])

**Comment by nikhilreddy06228** [comment]
lab06
